openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,133 @@
+package managementcpusoverride
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"k8s.io/apiserver/pkg/admission"
+	coreapi "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/kubelet/cm"
+)
+
+const (
+	PluginName                 = "autoscaling.openshift.io/ManagementCPUsOverride"
+	podWorkloadManagementLabel = "workload.openshift.io/target=management"
+	containerCPUsAnnotation    = "io.openshift.workload.management.cpu"
+)
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(config io.Reader) (admission.Interface, error) {
+			return &managementCPUsOverride{
+				Handler: admission.NewHandler(admission.Create),
+			}, nil
+		})
+}
+
+type managementCPUsOverride struct {
+	*admission.Handler
+}
+
+var _ = admission.MutationInterface(&managementCPUsOverride{})
+
+func (a *managementCPUsOverride) Admit(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") || attr.GetSubresource() != "" {
+		return nil
+	}
+
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+
+	// TODO: probably we will need to refactor it once we will have more workload types supported
+	if _, ok := pod.Labels[podWorkloadManagementLabel]; !ok {
+		return nil
+	}
+
+	// the only relevant pod QoS class that we interested is Burstable one
+	// only relevant resources to decide the QoS class is CPU and memory
+	if !isBurstable(pod, []coreapi.ResourceName{coreapi.ResourceCPU, coreapi.ResourceMemory}) {
+		return nil
+	}
+
+	// we should not remove the CPU request if it will change the QoS class of the pod
+	// check if the pod will stay burstable once we remove the CPU request
+	if !isBurstable(pod, []coreapi.ResourceName{coreapi.ResourceMemory}) {
+		return nil
+	}
+
+	updateContainerResources(pod)
+	return nil
+}
+
+func updateContainerResources(pod *coreapi.Pod) {
+	if pod.Annotations == nil {
+		pod.Annotations = map[string]string{}
+	}
+
+	for i, c := range pod.Spec.InitContainers {
+		if c.Resources.Requests != nil {
+			if _, ok := c.Resources.Requests[coreapi.ResourceCPU]; !ok {
+				continue
+			}
+
+			cpuRequest := c.Resources.Requests[coreapi.ResourceCPU]
+			cpusAnnotationKey := fmt.Sprintf("%s/%s", containerCPUsAnnotation, c.Name)
+			cpuShares := cm.MilliCPUToShares(cpuRequest.MilliValue())
+			pod.Annotations[cpusAnnotationKey] = fmt.Sprintf("%d", cpuShares)
```

Comment:
I thought we were going to rewrite the resource to a new type of resource as opposed to inventing an annotation.

Reply:
Yes, we should add a new resource, but I wanted to wait for the kubelet PR to be merged, otherwise the pod will stuck in the pending state forever and we still want to have annotation, because the CRI-O does not have any information regarding the pod spec

Reply:
We still need the annotation for cpu shares since the resource doesn't get passed down to CRI-O.